### PR TITLE
feat: Add fully configurable composer ignore platform requirement configuration

### PR DESCRIPTION
**Pull request by bobvandevijver**
## Changes:

This PR introduces the `composerIgnorePlatformReq` configuration options, which can be used to selectively ignore certain php platform requirement (when using Composer) instead of the current default that ignores all. 

## Context:

With #11025 I was actually hoping that wrongly updated packages wouldn't occur anymore as Composer would now run on the correct PHP version, but unfortunately this was not the case. Basically, this comment (https://github.com/renovatebot/renovate/issues/2355#issuecomment-830911434) was still happening.

But, as Renovate now uses the correct PHP version, only the extensions should be ignored. This change allows to do so.

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: docs/usage/self-hosted-configuration.md
Comment (viceice) [inline docs/usage/self-hosted-configuration.md:133]: Can't we combine those options to one ?

Comment (bobvandevijver) [reply docs/usage/self-hosted-configuration.md:133]: You mean like making the plural option accept both a bool and an array of strings?

Comment (viceice) [reply docs/usage/self-hosted-configuration.md:133]: Yes. Not sure if it's currently supported.

Comment (bobvandevijver) [reply docs/usage/self-hosted-configuration.md:133]: I just checked, that is currently not supported (see https://github.com/renovatebot/renovate/blob/main/lib/config/types.ts#L350).

**Comment by bobvandevijver** [comment]
The failed check is caused by an irrelevant security error in an unchanged file: `The hard-coded value "Bearer abc" is used as authorization header.`. I don't think I can do anything to solve that.

File: docs/usage/self-hosted-configuration.md
Comment (viceice) [reply docs/usage/self-hosted-configuration.md:129]: New idea:

- migrate `composerIgnorePlatformReqs=false` to `composerIgnorePlatformReqs=null`.
- if `composerIgnorePlatformReqs=[]` add `--ignore-platform-reqs` to command
- if `composerIgnorePlatformReqs` is nonempty string array, add individual `--ignore-platform-req xxx` options

**Review by viceice — COMMENTED** [review]
@rarkins WDYT about my idea above?

File: docs/usage/self-hosted-configuration.md
Comment (bobvandevijver) [reply docs/usage/self-hosted-configuration.md:129]: That actually sounds like a great solution! I will however wait for @rarkins to share his opinion before updating the implementation 😄

Comment (bobvandevijver) [reply docs/usage/self-hosted-configuration.md]: I had an empty moment, so I implemented it anyways. It already includes to configuration migration, so it should be ready to go 👍🏻

File: lib/config/migration.spec.ts
Comment (viceice) [inline lib/config/migration.spec.ts:703]:
```suggestion

  it('it migrates composerIgnorePlatformReqs values', () => {
```

readabillity

**Review by viceice — DISMISSED** [review]

**Review by viceice — DISMISSED** [review]

File: docs/usage/self-hosted-configuration.md
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:131]:
```suggestion
Set the platform requirements you want to exclude when using Composer.
Each item will be added to the Composer command with `--ignore-platform-req`.
When set to an empty array, `--ignore-platform-reqs` will be added to the Composer command.
```

We use one sentence per line. 😉 

I'd like an explanation that covers _why_ you might want to ignore/force through some platform requirements when you use Composer in a certain way.

File: docs/usage/self-hosted-configuration.md
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:133]: This line feels incomplete.

First we say: set this to null to prevent usage of ..., and it seems you also need to do something when invoking `composer` on the command line (`--ignore-platform-req`) Are we sure this is clear enough for Composer users?

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: docs/usage/self-hosted-configuration.md
Comment (bobvandevijver) [reply docs/usage/self-hosted-configuration.md]: Done!

File: docs/usage/self-hosted-configuration.md
Comment (bobvandevijver) [reply docs/usage/self-hosted-configuration.md]: I believe Composer users should know about the platform requirements and how to ignore them, but I updated the text to be more explicit.

File: docs/usage/self-hosted-configuration.md
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:139]: @rarkins I think we want to get rid of using `null`? And instead more descriptive terms like `auto/disabled/enabled`? 

- https://github.com/renovatebot/renovate/issues/9100

Do we want to replace the current `null` with `auto` maybe?

File: docs/usage/self-hosted-configuration.md
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:140]: How can you see/know if your "Renovate image is fully compatible with the Composer platform requirements"?

**Review by HonkingGoose — COMMENTED** [review]
I like that we now have a way better explanation on why you might need this option, well done!

I still have some comments. One is for the maintainer/author of Renovate, and one is for you. 😉

File: docs/usage/self-hosted-configuration.md
Comment (bobvandevijver) [reply docs/usage/self-hosted-configuration.md:139]: In my opinion `auto` here should mean that renovate automatically extracts all platform requirements from the project configuration to add them to the ignore requirements (except for PHP itself), but that would be hard to accomplish as not all requirements are available just from the repository (dependencies can also configure certain requirements). See also https://github.com/renovatebot/renovate/issues/2355#issuecomment-849431921.

Secondly, I do not believe combining textual options with a String option is currently supported by the configuration definition: https://github.com/renovatebot/renovate/blob/main/lib/config/types.ts#L350.

File: docs/usage/self-hosted-configuration.md
Comment (bobvandevijver) [reply docs/usage/self-hosted-configuration.md:140]: Trial and error I guess, you will see the problem soon enough as either the Composer invocation will fail completely, or you will see dependencies being updated that won't work in your other pipelines (one of the errors I currently see during my validation pipelines):

```
$ composer install --no-interaction --no-progress
Installing dependencies from lock file (including require-dev)
Verifying lock file contents can be installed on current platform.
Your lock file does not contain a compatible set of packages. Please run composer update.
  Problem 1
    - psr/cache is locked to version 2.0.0 and an update of this package was not requested.
    - psr/cache 2.0.0 requires php >=8.0.0 -> your php version (7.3.27) does not satisfy that requirement.
  Problem 2
    - psr/cache 2.0.0 requires php >=8.0.0 -> your php version (7.3.27) does not satisfy that requirement.
    - doctrine/annotations 1.13.2 requires psr/cache ^1 || ^2 || ^3 -> satisfiable by psr/cache[2.0.0].
    - doctrine/annotations is locked to version 1.13.2 and an update of this package was not requested.
```

I would actually strongly recommend against using the `null` possibility for normal usage.

File: lib/config/options/index.ts
Comment (bobvandevijver) [inline lib/config/options/index.ts:304]: @viceice I just saw that with my latest master merge this option (didn't know what its purpose was before, as it was `admin`) this option should be disabled in my opinion as you would want to be able to configure the excluded composer requirements on repository level. Is there anything else I need to adjust?

File: docs/usage/self-hosted-configuration.md
Comment (viceice) [reply docs/usage/self-hosted-configuration.md]: @HonkingGoose This is a string array option now. so #9100 did not apply here.

File: lib/manager/composer/artifacts.ts
Comment (viceice) [inline lib/manager/composer/artifacts.ts:141]:
```suggestion
          args += ' --ignore-platform-req ' + quote(req);
```

We should add proper quotes to escape improper / malcious values, needs `import { quote } from 'shlex';`

File: lib/config/options/index.ts
Comment (viceice) [reply lib/config/options/index.ts:304]: This is / was a global (previously admin) option, so can only be set by bot admins.

@rarkins Can we open this option for repo config? I don't see a huge risk here, needs proper quotes as i suggested below

**Review by viceice — CHANGES_REQUESTED** [review]

File: docs/usage/self-hosted-configuration.md
Comment (HonkingGoose) [reply docs/usage/self-hosted-configuration.md]: Oh okay. 😄 Sorry for the confusion!

File: lib/manager/composer/artifacts.ts
Comment (bobvandevijver) [reply lib/manager/composer/artifacts.ts]: Done!

**Review by viceice — DISMISSED** [review]

File: lib/config/options/index.ts
Comment (rarkins) [reply lib/config/options/index.ts:304]: OK

File: docs/usage/self-hosted-configuration.md
Comment (bobvandevijver) [reply docs/usage/self-hosted-configuration.md]: @HonkingGoose Would like to see an adjustment here, or is it fine as it is now?

Comment (HonkingGoose) [reply docs/usage/self-hosted-configuration.md]: If you recommend against using `null` in normal cases, we should definitely mention that.

Can you add a short explanation in the docs? I think what you had posted above can be edited to fit in?

Comment (bobvandevijver) [reply docs/usage/self-hosted-configuration.md]: Okay, I did my best, I really believe this should be clear enough for any Composer user, otherwise I do not think they should bother using Renovate 😄

**Review by HonkingGoose — DISMISSED** [review]
Docs look way better now! Thank you for adding the new text!

I'll let the maintainers review the code itself.

**Review by viceice — APPROVED** [review]

**Comment by mbrodala** [comment]
It should be noted that this requires Composer 2 which introduced the `--ignore-platform-req` argument.

Also from what I can see this is a breaking change since the global `composerIgnorePlatformReqs` option (and `RENOVATE_COMPOSER_IGNORE_PLATFORM_REQS` environment variable) does not work anymore with this. I'm not sure if Renovate follows semver though. If yes, then this should have been released as version 30.x.

Instead one must now add `composerIgnorePlatformReqs` to each single repository config (lucky if you use a shared config) which can be quite a burden depending on the amount of repositories.

Am I wrong?

**Comment by viceice** [comment]
Repo config can also be used globally, so no breaking.

Don't we have added a migration from env config?

If we have this, it won't be a breaking change

**Comment by bobvandevijver** [comment]
> It should be noted that this requires Composer 2 which introduced the --ignore-platform-req argument.

This a very good note, thank you (updated docs with https://github.com/renovatebot/renovate/pull/11399). 

Note that by default this argument is not used, only when you explicitly configure the extensions. The default behaviour is the same.

**Comment by mbrodala** [comment]
> Repo config can also be used globally, so no breaking.

That's good to know. Is that actually mentioned anywhere? I cannot find anything about this on [Configuration Options](https://docs.renovatebot.com/configuration-options/) or [Self-Hosted configuration options](https://docs.renovatebot.com/self-hosted-configuration/)

> Don't we have added a migration from env config?

Yeah, I just saw that the option is generally migrated from the former boolean values.

> If we have this, it won't be a breaking change

Agreed. :-)

> This a very good note, thank you (updated docs with #11399).

Thanks. :+1:

**Comment by viceice** [comment]
> > Repo config can also be used globally, so no breaking.
> 
> That's good to know. Is that actually mentioned anywhere? I cannot find anything about this on [Configuration Options](https://docs.renovatebot.com/configuration-options/) or [Self-Hosted configuration options](https://docs.renovatebot.com/self-hosted-configuration/)

We don't recommend to configure repo options globally, as normal users are not aware of those.
So best is to create a shared preset for general options. 

> 
> > Don't we have added a migration from env config?
> 
> Yeah, I just saw that the option is generally migrated from the former boolean values.
> 

Any option, which don't have `env: false` can be configured via env variable.

**Comment by mbrodala** [comment]
> We don't recommend to configure repo options globally, as normal users are not aware of those.
> So best is to create a shared preset for general options.

That's actually a good idea. We recently introduced an organization-level config and with this change we can move the option to that config.

**Pull request merged**
